### PR TITLE
enable asyncio using pio

### PR DESCRIPTION
**Pull request by jedwards4b**
### Description of changes 
Allows IO tasks to be independent of compute tasks in cesm

### Specific notes
(testing in progress)
Contributors other than yourself, if any:
Depends on share (https://github.com/ESCOMP/CESM_share/pull/37) and cime (https://github.com/ESMCI/cime/pull/4340).  

CMEPS Issues Fixed (include github issue #):

Are changes expected to change answers? (specify if bfb, different at roundoff, more substantial) 

Any User Interface Changes (namelist or namelist defaults changes)?

### Testing performed

Testing performed if application target is CESM:
- [ ] (recommended) CIME_DRIVER=nuopc scripts_regression_tests.py
   - machines:
   - details (e.g. failed tests):
- [ ] (recommended) CESM testlist_drv.xml
   - machines and compilers:
   - details (e.g. failed tests):
- [X] (optional) CESM prealpha test
   - machines and compilers cheyenne intel 
   - details (e.g. failed tests): results consistant with cesm2_3_alpha10d
- [ ] (other) please described in detail
   - machines and compilers
   - details (e.g. failed tests):

Testing performed if application target is UFS-coupled:
- [ ] (recommended) UFS-coupled testing
   - description:
   - details (e.g. failed tests):

Testing performed if application target is UFS-HAFS:
- [X] (recommended) UFS-HAFS testing
   - description:
   - details (e.g. failed tests):

### Hashes used for testing:

- [ ] CESM:
  - repository to check out: https://github.com/ESCOMP/CESM.git
  - branch/hash:
- [ ] UFS-coupled, then umbrella repostiory to check out and associated hash:
  - repository to check out:
  - branch/hash:
- [ ] UFS-HAFS, then umbrella repostiory to check out and associated hash:
  - repository to check out:
  - branch/hash:

### CODE REVIEW

**Review by billsacks — CHANGES_REQUESTED** [review]
Thanks a lot for your work on this @jedwards4b ! This is a really exciting new feature!

I have a few comments below. Most of them are minor & easy. The main one that might be more involved is my long comment, particularly the second paragraph starting "Part of what I'm wondering about here...". Let me know if you'd like to talk about this.

File: cesm/driver/ensemble_driver.F90 [hunk 1]
```diff
@@ -53,6 +62,14 @@ subroutine SetServices(ensemble_driver, rc)
          specRoutine=SetModelServices, rc=rc)
     if (chkerr(rc,__LINE__,u_FILE_u)) return
 
+    ! ModifyCplLists is a NUOPC specialization which happens after Advertize but before Realize
```
[the remark this answers is not in the document]
Comment (jedwards4b) [reply cesm/driver/ensemble_driver.F90:65]: done

File: cesm/driver/ensemble_driver.F90 [hunk 2]
```diff
@@ -203,23 +259,32 @@ subroutine SetModelServices(ensemble_driver, rc)
...
+       do n=1,ntasks_per_member+pio_asyncio_ntasks
+          if(pio_asyncio_stride == 0 .or. modulo(n,pio_asyncio_rootpe+1) .ne. 0) then
+             petList(petcnt) = currentpet
+             petcnt = petcnt+1
+             if (currentpet == localPet) comp_task=.true.
+          else
+             asyncio_petlist(iopetcnt) = currentpet
+             iopetcnt = iopetcnt + 1
+             if (currentpet == localPet) asyncio_task=.true.
+          endif
```
[the remark this answers is not in the document]
Comment (jedwards4b) [reply cesm/driver/ensemble_driver.F90:278]: I think that you may be right here.  I'll create a few scenarios to test this and put in a fix if needed.

File: cesm/driver/esm.F90 [hunk 3]
```diff
@@ -932,8 +931,8 @@ subroutine esm_init_pelayout(driver, maxthreads, rc)
 
     ! Initialize PIO
     ! This reads in the pio parameters that are independent of component
-    call driver_pio_init(driver, rc=rc)
-    if (chkerr(rc,__LINE__,u_FILE_u)) return
+!    call driver_pio_init(driver, rc=rc)
+!    if (chkerr(rc,__LINE__,u_FILE_u)) return
```
[the remark this answers is not in the document]
Comment (jedwards4b) [reply cesm/driver/esm.F90:935]: done

File: cesm/driver/esm.F90 [hunk 4]
```diff
@@ -1180,8 +1179,8 @@ subroutine esm_init_pelayout(driver, maxthreads, rc)
     enddo
     ! Read in component dependent PIO parameters and initialize
     ! IO systems
-    call driver_pio_component_init(driver, size(comps), rc)
-    if (chkerr(rc,__LINE__,u_FILE_u)) return
+!    call driver_pio_component_init(driver, size(comps), rc)
+!    if (chkerr(rc,__LINE__,u_FILE_u)) return
```
[the remark this answers is not in the document]
Comment (jedwards4b) [reply cesm/driver/esm.F90:1183]: done

File: cesm/driver/esm_time_mod.F90 [hunk 5]
```diff
@@ -101,99 +102,162 @@ subroutine esm_time_clockInit(ensemble_driver, instance_driver, logunit, mastert
...
+    if (ChkErr(rc,__LINE__,u_FILE_u)) return
+    read(cvalue,*) rof_cpl_dt
+
+    call NUOPC_CompAttributeGet(ensemble_driver, name="wav_cpl_dt", value=cvalue, rc=rc)
+    if (ChkErr(rc,__LINE__,u_FILE_u)) return
+    read(cvalue,*) wav_cpl_dt
+
+    call NUOPC_CompAttributeGet(ensemble_driver, name="glc_avg_period", value=glc_avg_period, rc=rc)
+    if (ChkErr(rc,__LINE__,u_FILE_u)) return
+    read(cvalue,*) glc_avg_period
```
[the remark this answers is not in the document]
Comment (jedwards4b) [reply cesm/driver/esm_time_mod.F90:160]: done

File: cesm/nuopc_cap_share/driver_pio_mod.F90 [hunk 6]
```diff
@@ -247,84 +314,198 @@ subroutine driver_pio_component_init(driver, ncomps, rc)
...
              read(cval, *) pio_comp_settings(i)%pio_root
-             
+          
              if(pio_comp_settings(i)%pio_root < 0 .or. pio_comp_settings(i)%pio_root > npets) then
                 pio_comp_settings(i)%pio_root = 0
              endif
+          endif
           
-             call NUOPC_CompAttributeGet(gcomp(i), name="pio_typename", value=cval, rc=rc)
+          call NUOPC_CompAttributeGet(gcomp(i), name="pio_async_interface", value=cval, rc=rc)
```
[the remark this answers is not in the document]
Comment (jedwards4b) [reply cesm/nuopc_cap_share/driver_pio_mod.F90:327]: thanks for spotting that.

File: cesm/driver/ensemble_driver.F90 (hunk 2 above)
[the remark this answers is not in the document]
Comment (jedwards4b) [reply cesm/driver/ensemble_driver.F90:278]: It's difficult to get to the pio_async_interface flags at this point in the fortran and I am thinking about implementing something in the buildnml instead - would that be acceptable?

File: mediator/med.F90 [hunk 7]
```diff
@@ -59,10 +59,10 @@ module MED
   public  SetServices
   public  SetVM
   private InitializeP0
-  private InitializeIPDv03p1 ! advertise fields
-  private InitializeIPDv03p3 ! realize connected Fields with transfer action "provide"
-  private InitializeIPDv03p4 ! optionally modify the decomp/distr of transferred Grid/Mesh
-  private InitializeIPDv03p5 ! realize all Fields with transfer action "accept"
+  private AdvertiseFields ! advertise fields
```
Comment (DeniseWorthen) [inline mediator/med.F90:62]: Thanks for these.

**Review by DeniseWorthen — APPROVED** [review]
All UWM test pass.

**Review by uturuncoglu — APPROVED** [review]

**Comment by jedwards4b** [comment]
Retested prealpha against alpha12b
```
 ./cs.status.20230125_114302_bxq3t5 -f | grep -v NLCOMP | grep -v BFAIL
20230125_114302_bxq3t5: 87 tests
    FAIL DAE_N2_D_Lh12_Vmct.f10_f10_mg37.I2000Clm50BgcCrop.cheyenne_intel.clm-DA_multidrv TPUTCOMP Error: TPUTCOMP: Computation time increase > 25% from baseline
    FAIL ERP_D_Ln9_Vnuopc.C48_C48_mg17.QPC6.cheyenne_intel.cam-outfrq9s MODEL_BUILD time=245
    FAIL ERP_Ld3_Vnuopc.f09_f09_mg17.FCfireHIST.cheyenne_intel.cam-outfrq1d BASELINE cesm2_3_alpha12b: DIFF
    FAIL IRT_C3_Ld7.f19_g17.BHIST.cheyenne_intel.allactive-defaultio RUN time=7222
    PEND IRT_C3_Ld7.f19_g17.BHIST.cheyenne_intel.allactive-defaultio COMPARE_base_restart
    FAIL MCC.f19_g17.B1850.cheyenne_intel.allactive-defaultiomi RUN time=7231
    PEND MCC.f19_g17.B1850.cheyenne_intel.allactive-defaultiomi COMPARE_base_single_instance
    FAIL MCC_Ld5.f19_g17.B1850G.cheyenne_intel.allactive-cism-test_coupling RUN time=5427
    PEND MCC_Ld5.f19_g17.B1850G.cheyenne_intel.allactive-cism-test_coupling COMPARE_base_single_instance
    FAIL SMS_D_Ld1_PS.f09_g17.I1850Clm50BgcSpinup.cheyenne_intel.clm-cplhist RUN time=46
    FAIL SMS_D_Ln9_Vnuopc_P720x1.ne0CONUSne30x8_ne0CONUSne30x8_mt12.FCnudged.cheyenne_intel.cam-outfrq9s RUN time=1216
    FAIL SMS_Ld1_P144_D.T62_g17.C.cheyenne_intel.pop-144blocks_320x384_spacecurve RUN time=43244
    FAIL SMS_Ld2_P80_D.T62_g37.C1850ECO.cheyenne_intel.pop-ecosys_81blocks_100x116_spacecurve RUN time=43257
```
@fischer-ncar these all look like failures in alpha12b - can I go ahead and merge?

**Comment by fischer-ncar** [comment]
Yes, those are expected failures.  You can go ahead and merge.

**Pull request merged**